 #!/bin/sh
                #PBS -l walltime=24:00:00
                #PBS -l cput=24:00:00
                #PBS -l nodes=1:ppn=4:bio7
                #PBS -m abe
                #PBS -M 2196709r@student.gla.ac.uk
                #PBS -N Kiron_projn
                #PBS -q bioinf-stud
                #PBS -d /export/home/biostuds/2196709r/projFQ

#Storing file paths to required files as variables  
R1adapt='/export/home/biostuds/2196709r/projFQ/R1adapt.fasta'
R2adapt='/export/home/biostuds/2196709r/projFQ/R2adapt.fasta'
reference='/export/home/biostuds/2196709r/projFQ/gencode.v34.annotation.gtf'

#Loop through each folder, where each directory contains a pair of reads
for file in n1 n2 n3 n4 n5 n6 n7 n8
do

#Enter the directory, store the contents of the directory as a list, and create a variable for forward and reverse reads 
cd $file
files="$(ls -a)"
set -- junk $files
shift

#Creates variables for the various file formats required, basing the names of the variables upon the stored contents of the directory
f1a=$3
f1b=$4
f1a=${f1a%.fastq.gz}
f1b=${f1b%.fastq.gz}
raw1="${f1a}.fastq.gz"
raw2="${f1b}.fastq.gz"
trim1a="${f1a}_tr1a.fastq"
trim1b="${f1b}_tr1b.fastq"
trim2a="${f1a}_tr2a.fastq"
trim2b="${f1b}_tr2b.fastq"
singles="singles.fastq"

#Fastqc of the raw data
fastqc $raw1
fastqc $raw2

#Pre-processing steps, adapter and quality trimming
scythe -o $trim1a -a $R1adapt -q sanger $raw1
scythe -o $trim1b -a $R2adapt -q sanger $raw2
sickle pe -f $trim1a -r $trim1b -t sanger -o $trim2a -p $trim2b -s $singles -q 20

#Fastqc of the pre-proccessed data
fastqc $trim1a
fastqc $trim1b
fastqc $singles

#Indexing reads
hisat2 --rna-strandness R -q -x /export/home/biostuds/2196709r/projFQ/grch38_tran/genome_tran -1 $trim2a -2 $trim2b -U $singles -S ${file}_hisat.sam

#converting from sam to bam, and sorting bam file
samtools view -bS -o ${file}.bam ${file}_hisat.sam
samtools sort ${file}.bam ${file}_sorted.bam 

#Assembling reads
stringtie ${file}_sorted.bam.bam --rf -p 4 -e -G $reference -B -o $file/${file}.gtf

#Removing large intermediate files
rm ${file}_sorted.bam.bam ${file}.bam ${file}_hisat.sam $singles $trim2b $trim2a $trim1b $trim1a

#Move up a directory
cd ..
done

#Path to output files from loop contained within sample_list.txt, these files are converted to gene/transcript count matrices by prepDE
python2.7 /export/projects/polyomics/App/prepDE.py -i /export/home/biostuds/2196709r/projFQ/sample_list.txt 

