# Affymetrix microarray differential expression 
# The data analyzed here are a clinical microarray dataset
# of Chronic Myeloid Leukaemia (CML) Leukaemic stem cells (LSCs)
# or Progenitors vs their non-leukemia counterparts

my_theme <- theme(
  panel.grid = element_blank(),
  panel.background = element_blank(),
  panel.border = element_rect(colour = "black", fill=NA, size=2),
  plot.background = element_blank(),
  legend.background = element_rect(fill="transparent", colour=NA),
  legend.key = element_rect(fill="transparent", colour=NA),
  plot.title = element_text(size = 33, margin = margin(b = 5),hjust=0.5,vjust=0.5, family="Arial", face="bold"),
  title = element_text(size = 16, margin = margin(b = 5),hjust=0,vjust=0.5, family="Arial", face="bold"),
  axis.text.y = element_text(size = 14, margin = margin(r = 5),hjust=1,vjust=0.5, family="Arial", face="bold",colour="black"),
  axis.text.x = element_text(size = 14, margin = margin(t = 5),hjust=0.5,vjust=1, family="Arial", face="bold",colour="black"),
  axis.title.y = element_text(size = 16, margin = margin(r = 10),angle = 90,hjust=0.5,vjust=0.5, family="Arial", face="bold"),
  axis.title.x = element_text(size = 16, margin = margin(t = 10),hjust=0.5,vjust=1, family="Arial", face="bold"),
  legend.text=element_text(size=16, family="Arial", face="bold"),
  legend.title=element_blank(),
  legend.key.size=unit(1,"line"),
  plot.margin=unit(c(1,1,1,1), "cm"),
  strip.text.x = element_text(size = 16, family="Arial", face="bold", vjust=1),
  strip.background = element_rect(fill="transparent", size=0),
  panel.spacing = unit(1, "lines")
)

# list of packages used

#General Bioconductor packages
library(Biobase)
library(oligoClasses)

#Annotation and data import packages
library(ArrayExpress)
library(pd.hugene.1.0.st.v1)
library(hugene10sttranscriptcluster.db)

#Quality control and pre-processing packages
library(oligo)
library(arrayQualityMetrics)

#Analysis and statistics packages
library(limma)
#library(topGO)
#library(ReactomePA)
library(clusterProfiler)

#Plotting and color options packages
library(gplots)
library(ggplot2)
library(geneplotter)
library(RColorBrewer)
library(pheatmap)
library(enrichplot)

#Formatting/documentation packages
#library(rmarkdown)
#library(BiocStyle)
library(dplyr)
library(tidyr)

#Helpers:
library(stringr)
library(matrixStats)
library(genefilter)
library(openxlsx)
#library(devtools)

## raw data CEL files download
# the CEL are produced by the array scanner and describe the 
# probe intensities. The data are deposited at ArrayExpress

# anno_AE <- getAE("E-MTAB-2581", type = "raw") # run just once

# the aim is to create an ExpressionSet
# Investigation Description Format (IDF): info about experiment
# Sample and Data Relationship Format (SDRF): info about samples

# import SDRF to get samples info

sdrf_location <- file.path("E-MTAB-2581.sdrf.txt")
SDRF <- read.delim(sdrf_location) # import the SDRF from raw data folder

rownames(SDRF) <- SDRF$Array.Data.File
SDRF <- AnnotatedDataFrame(SDRF) # turn SDRF into AnnotatedDataFrame which will be used to create the ExpressionSet

# create the ExpressionSet
# import the CEL files into the raw_data variable using read.celfiles()
# it is important to import the CEL files in the order that corresponds to
# the SDRF table, thus filenames argument has the $ symbol

raw_data <- read.celfiles(filenames = SDRF$Array.Data.File,
                          verbose = FALSE,
                          phenoData = SDRF)

head(pData(raw_data))
sample_info <- pData(raw_data) # to retrieve sample information

sample_info$group <- c("LSC", "LSC", "LSC", "LSC", "LSC", "LSC",
                       "LPC", "LPC", "LPC", "LPC", "LPC", "LPC",
                       "HSC", "HSC", "HSC", "HSC", "HSC", "HSC",
                       "HPC", "HPC", "HPC", "HPC", "HPC", "HPC")

sample_info$group <- factor(sample_info$group, levels = c("HPC", "HSC", "LSC", "LPC"))

### Quality control

em <- exprs(raw_data) # retrieve the expression matrix with probes intensities
# log2 transformation
em_log2 <- log2(em)

# density plot of non-normalized raw data
plotDensities(em, group = sample_info[,"group"],legend = "topright")
# density plot of log2 transformed non-normalized data
plotDensities(em_log2, group = sample_info[,"group"],legend = "topright")




# PCA
pca <- prcomp(t(em_log2), scale. = FALSE)
percentVar <- round(100*pca$sdev^2/sum(pca$sdev^2),1)
sd_ratio <- sqrt(percentVar[2]/percentVar[1])

# create data frame for PCA plot
dataGG <- data.frame(PC1 = pca$x[,1],
                     PC2 = pca$x[,2],
                     Disease = pData(raw_data)$Factor.Value.disease.,
                     Cell_Type = pData(raw_data)$Factor.Value.cell.type.,
                     Sample_ID = pData(raw_data)$Source.Name)

ggplot(dataGG, aes(x = PC1, y = PC2)) +
  geom_point(aes(color = sample_info$group), size = 2) +
  ggtitle("PCA plot of log2-transformed raw data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  my_theme +
  scale_color_manual(values = c("limegreen", "darkmagenta", "turquoise4", "royalblue4"))

# the PCA plot shows that both Disease and Cell_Type drive different transcription profiles. the source of variation should be DIsease only

# boxplot of raw_data
# boxplot() from oligo package does the log2-transformation by default, hence the use of raw_data

oligo::boxplot(raw_data, target = "core", main = "Boxplot of log2-intensities for the raw data") # need normalization

## Backgroud adjustement
# get rid of non-specific hybridization and noise

## Across-array normalization (calibration)
# to compare measurements from different array hybridizations and get rid of sources of variation

## Summarization
# on the Affymetrics platform transcripts are represented by multiple probes
# so it is important to summarize the normalized probes into one quantity that is proportional to the 
# amount of the RNA transcript

# can be done in one step using the oligo RMA (robust multichip average) algorithm for summarization

## Relative Log Expression (RLE)
# this is a further quality control. It calculates the median log2 intensity of every transcript across all arrays
# then the transcript median is subtrancted from every transcript intensity via sweep()

rle_eset <- rma(raw_data, target = "core", normalize = FALSE) # RMA without the normalization
row_medians_assayData <- Biobase::rowMedians(as.matrix(exprs(rle_eset)))
rle_data <- sweep(exprs(rle_eset), 1, row_medians_assayData)
rle_data <- as.data.frame(rle_data)
rle_data_gathered <- gather(rle_data, patient_array, log2_expression_deviation)

ggplot(rle_data_gathered, aes(x = patient_array, y = log2_expression_deviation)) +
  geom_boxplot(outlier.shape = NA) +
  ylim(c(-2,2)) +
  theme(axis.text.x = element_text(color = "aquamarine4",
                                    angle = 60, size = 6.5, hjust = 1, face = "bold"))


# RMA calibration
raw_data_normalized <- rma(raw_data, target = "core")
## Quality assessment of the calibated data
# PCA

em_normalized <- exprs(raw_data_normalized)
pca_normalized <- prcomp(t(em_normalized), scale. = FALSE)
percentVar_normalized <- round(100*pca_normalized$sdev^2/sum(pca_normalized$sdev^2),1)
sd_ratio_normalized <- sqrt(percentVar_normalized[2]/ percentVar_normalized[1])

dataGG_normalized <- data.frame(PC1 = pca_normalized$x[,1], PC2 = pca_normalized$x[,2],
                                Disease = pData(raw_data_normalized)$Factor.Value.disease.,
                                Cell_Type = pData(raw_data_normalized)$Factor.Value.cell.type.,
                                Sample_ID = pData(raw_data_normalized)$Source.Name)

ggplot(dataGG_normalized, aes(x = PC1, y = PC2, color = sample_info$group)) +
  geom_point(size = 3, alpha = 0.6) +
  ggtitle("PCA") +
  xlab(paste0("PC1 (", percentVar_normalized[1], "%)")) +
  ylab(paste0("PC2 (", percentVar_normalized[2], "%)")) +
  my_theme +
  scale_color_manual(values = c("limegreen", "darkmagenta", "turquoise4", "royalblue4"))

# this new PCA plot shows that the source of variation is due to disease.
# differential expression between disease is the main source of variation.

oligo::boxplot(raw_data_normalized, target = "core", main = "Boxplot of log2-intensities for the normalized raw data") # data are normalized now



# clustering heatmap

cell_type_names <- ifelse(str_detect(pData(raw_data_normalized)$Factor.Value.cell.type.,
                                     "progenitor"), "Progenitor", "Stem cell")
disease_names <- ifelse(str_detect(pData(raw_data_normalized)$Factor.Value.disease.,
                                   "leukemia"), "CML", "Normal")
annotation_for_heatmap <- data.frame(Cell_type = cell_type_names, Disease = disease_names, Group = sample_info$group)
row.names(annotation_for_heatmap) <- row.names(pData(raw_data_normalized))

dists <- as.matrix(dist(t(em_normalized), method = "manhattan")) # compute distances with dist() 
rownames(dists) <- row.names(pData(raw_data_normalized))
hmcol <- rev(colorRampPalette(brewer.pal(9, "YlOrRd"))(255))
diag(dists) <- NA

ann_colors <- list(
  #Cell_type = c(Progenitor = "chartreuse4", Stem_cell = "burlywood3"),
  Disease = c(CML = "black", Normal = "turquoise"),
  Group = c(HSC = "darkmagenta", LSC = "turquoise4", LPC = "royalblue4", HPC = "limegreen")
  )

pheatmap(dists, col = (hmcol), 
         annotation_row = annotation_for_heatmap,
         annotation_colors = ann_colors,
         legend = TRUE, 
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                           max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "Clustering heatmap")

# samples cluster strongly by Cell_type, although not perfectly.


# filtering based on intensity
# filter out lowly expressed genes

medians <- rowMedians(em_normalized)

# histogram of median intensities
hist(em_normalized, 100, col = "cornsilk1", freq = FALSE,
     main = "Histogram of the median intensities",
     border = "antiquewhite4",
     xlab = "Median intensities")

# alternative way of plotting the intensities
plotDensities(em_normalized, group = sample_info[,"group"],legend = "topright")


summary(em_normalized)

cut_off <- 5                        # cut-off for the min value
abline(v = cut_off, col = "coral4", lwd = 2)

# transcripts without intensities larger than the threshold in at least as many arrays as the smallest experimental group are excluded

no_of_samples <- 
  table(paste0(pData(raw_data_normalized)$Factor.Value.disease., "_", 
               pData(raw_data_normalized)$Factor.Value.cell.type.))
no_of_samples 
samples_cutoff <- min(no_of_samples) # min number of replicates per sample group

idx_threshold <- apply(em_normalized, 1, 
                       function(x){
                         sum(x > cut_off) >= samples_cutoff}) # applies the idx_threshold function to each row and excludes the ones that are not in at least 6 replicates
                         
table(idx_threshold)
raw_data_filtered <- subset(raw_data_normalized,idx_threshold)
em_filtered <- subset(em_normalized, idx_threshold)
plotDensities(em_filtered, group = sample_info[,"group"],legend = "topright")

pca_filtered <- prcomp(t(em_filtered), scale. = FALSE)
percentVar_filtered <- round(100*pca_filtered$sdev^2/sum(pca_filtered$sdev^2),1)
sd_ratio_filtered <- sqrt(percentVar_filtered[2]/ percentVar_filtered[1])

dataGG_filtered <- data.frame(PC1 = pca_filtered$x[,1], PC2 = pca_filtered$x[,2],
                                Disease = pData(raw_data_filtered)$Factor.Value.disease.,
                                Cell_Type = pData(raw_data_filtered)$Factor.Value.cell.type.,
                                Sample_ID = pData(raw_data_filtered)$Source.Name)

ggplot(dataGG_filtered, aes(x = PC1, y = PC2, color = sample_info$group)) +
  geom_point(size = 3, alpha = 0.6) +
  ggtitle("PCA") +
  xlab(paste0("PC1 (", percentVar_normalized[1], "%)")) +
  ylab(paste0("PC2 (", percentVar_normalized[2], "%)")) +
  my_theme +
  scale_color_manual(values = c("limegreen", "darkmagenta", "turquoise4", "royalblue4"))


## Retrieve annotations
# add the feauture data (i.e. annotations) to the filtered ExpressionSet

annotation_raw_data <- AnnotationDbi::select(hugene10sttranscriptcluster.db,
                                       keys = (featureNames(raw_data_normalized)),
                                       columns = c("SYMBOL", "GENENAME"),
                                       keytype = "PROBEID")

# select() from Annotation Dbi to retrieve gene symbols and a short description for each probe

annotations <- subset(annotation_raw_data, !is.na(SYMBOL)) # get rid of all the NAs for SYMBOLs

## remove multiple mappings
# many gens will map to multiple gene symbols

anno_grouped <- group_by(annotations, PROBEID)
anno_summarized <- summarize(anno_grouped, no_of_matches = n_distinct(SYMBOL))

head(anno_summarized)
anno_filtered <- filter(anno_summarized, no_of_matches >1)
head(anno_filtered)
probe_stats <- anno_filtered 

nrow(probe_stats)

ids_to_exlude <- (featureNames(raw_data_filtered) %in% probe_stats$PROBEID)
table(ids_to_exlude)

raw_data_final <- subset(raw_data_filtered, !ids_to_exlude)
validObject(raw_data_final)

head(annotations)

fData(raw_data_final)$PROBEID <- rownames(fData(raw_data_final))
fData(raw_data_final) <- left_join(fData(raw_data_final), annotations)

rownames(fData(raw_data_final)) <- fData(raw_data_final)$PROBEID 
validObject(raw_data_final)

annotations_final <- fData(raw_data_final)
annotations_final <- na.omit(annotations_final)

 
# differentially expressed genes

individual <- as.character(pData(raw_data_final)$Source.Name)

cell_type <- str_replace_all(pData(raw_data_final)$Factor.Value.cell.type., " ","_")
cell_type <- ifelse(cell_type == "hematopoietic_stem_cell", "SC", "PC")

disease <- str_replace_all(pData(raw_data_final)$Factor.Value.disease., " ", "_")
disease <- ifelse(disease == "chronic_myelogenous_leukemia", "L", "H")

## limma pipeline

group <- with(sample_info,                                             # with() to refer to the columns in samples dataframe
              paste(disease, cell_type, sep = "."))  

design <- model.matrix(~0 + group)
colnames(design) <- c("HPC", "HSC", "LPC", "LSC")

fit <- lmFit(em_filtered, design)  # fit linear model to samples

contrasts <- makeContrasts(LSC = "LSC - HSC",
                           LPC_gradual = "LPC - HSC",
                           HPC = "HPC - HSC",
                           LPC = "LPC - LSC",
                           LPC_net = "(LPC - LSC) - (LSC - HSC)",
                           LSC_vs_LPC = "LSC - LPC",
                           LSC_vs_LPC_net = "(LSC - HSC) - (LPC - HSC)",
                           levels = design)

fit2 <- contrasts.fit(fit, contrasts)
fit2 <- eBayes(fit2)

diff.exprs <- decideTests(fit2, p.value = 0.05, adjust.method = "fdr", lfc = 1) 
summary(diff.exprs)  # list of up/down regulated and non significant genes

# extract results

## Leukemic Stem Cells (LSC) ----

LSC <- topTable(fit2, coef = "LSC", number = Inf) # retrieve a data.frame of differential expression
hist(LSC$P.Value, col = brewer.pal(3, name = "Set2")[1], main = "Leukemic stem cells", xlab = "p-values")

LSC_annotated <- merge(LSC, annotations_final, by = 0)
row.names(LSC_annotated) <- LSC_annotated$Row.names

LSC_annotated <- LSC_annotated[, c(2,5,6,8,9,10)]

LSC_annotated$de <- "Non significant"
LSC_annotated$de[LSC_annotated$logFC > 1 & LSC_annotated$adj.P.Val < 0.05] <- "Upregulated"
LSC_annotated$de[LSC_annotated$logFC < -1 & LSC_annotated$adj.P.Val < 0.05] <- "Downregulated"

LSC_annotated <- arrange(LSC_annotated, -logFC)


rownames(LSC_annotated) <- make.names(LSC_annotated[,"SYMBOL"], unique = TRUE)

LSC_sig <- subset(LSC_annotated, adj.P.Val < 0.05 & abs(logFC)> 1.0) %>% arrange(-logFC)

## create a ranked list of genes for gsea

LSC_gsea <- arrange(LSC_annotated, -logFC)
rownames(LSC_gsea) <- make.names(LSC_gsea[,"SYMBOL"], unique = TRUE)
write.table(LSC_gsea, file="LSC_ranked.csv", sep = ",") # export file for GSEA software

# LSC vs LPC ----

LSC_vs_LPC <- topTable(fit2, coef = "LSC_vs_LPC", number = Inf) # retrieve a data.frame of differential expression

hist(LSC_vs_LPC$P.Value, col = brewer.pal(3, name = "Set2")[2], main = "LSC_vs_LPC", xlab = "p-values")

LSC_vs_LPC_annotated <- merge(LSC_vs_LPC, annotations_final, by = 0)
row.names(LSC_vs_LPC_annotated) <- LSC_vs_LPC_annotated$Row.names
LSC_vs_LPC_annotated <- LSC_vs_LPC_annotated[, c(2,5,6,8,9,10)]

LSC_vs_LPC_annotated$de <- "Non significant"
LSC_vs_LPC_annotated$de[LSC_vs_LPC_annotated$logFC > 1 & LSC_vs_LPC_annotated$adj.P.Val < 0.05] <- "Upregulated"
LSC_vs_LPC_annotated$de[LSC_vs_LPC_annotated$logFC < -1 & LSC_vs_LPC_annotated$adj.P.Val < 0.05] <- "Downregulated"

LSC_vs_LPC_annotated <- arrange(LSC_vs_LPC_annotated, -logFC)


rownames(LSC_vs_LPC_annotated) <- make.names(LSC_vs_LPC_annotated[,"SYMBOL"], unique = TRUE)

LSC_vs_LPC_sig <- subset(LSC_vs_LPC_annotated, adj.P.Val < 0.05 & abs(logFC)> 1.0) %>% arrange(-logFC)

## create a ranked list of genes for gsea

LSC_vs_LPC_gsea <- arrange(LSC_vs_LPC_annotated, -logFC)
rownames(LSC_vs_LPC_gsea) <- make.names(LSC_vs_LPC_gsea[,"SYMBOL"], unique = TRUE)
write.table(LSC_vs_LPC_gsea, file="LSC_vs_LPC_ranked.csv", sep = ",") # export file for GSEA software


# Hematopoietic Progenitor Cells (HPC) ----

HPC <- topTable(fit2, coef = "HPC", number = Inf) # retrieve a data.frame of differential expression

hist(HPC$P.Value, col = brewer.pal(3, name = "Set2")[3], main = "Hematopoietic progenitor cells", xlab = "p-values")

HPC_annotated <- merge(HPC, annotations_final, by = 0)
row.names(HPC_annotated) <- HPC_annotated$Row.names
HPC_annotated <- HPC_annotated[, c(2,5,6,8,9,10)]

HPC_annotated$de <- "Non significant"
HPC_annotated$de[HPC_annotated$logFC > 1 & HPC_annotated$adj.P.Val < 0.05] <- "Upregulated"
HPC_annotated$de[HPC_annotated$logFC < -1 & HPC_annotated$adj.P.Val < 0.05] <- "Downregulated"
rownames(HPC_annotated) <- make.names(HPC_annotated[,"SYMBOL"], unique = TRUE)


HPC_sig <- subset(HPC_annotated, adj.P.Val < 0.05 & abs(logFC)> 1.0) %>% arrange(-logFC)

HPC_gsea <- arrange(HPC_annotated, -logFC)
rownames(HPC_gsea) <- make.names(HPC_gsea[,"SYMBOL"], unique = TRUE)
write.table(HPC_gsea, file="HPC_ranked.csv", sep = ",") # export file for GSEA software

# LPC ----

LPC <- topTable(fit2, coef = "LPC", number = Inf) # retrieve a data.frame of differential expression

hist(LPC$P.Value, col = brewer.pal(3, name = "Set2")[4], main = "LPC_gradual", xlab = "p-values")

LPC_annotated <- merge(LPC, annotations_final, by= 0) 
row.names(LPC_annotated) <- LPC_annotated$Row.names
LPC_annotated <- LPC_annotated[, c(2,5,6,8,9,10)]

LPC_annotated$de <- "Non significant"
LPC_annotated$de[LPC_annotated$logFC > 1 & LPC_annotated$adj.P.Val < 0.05] <- "Upregulated"
LPC_annotated$de[LPC_annotated$logFC < -1 & LPC_annotated$adj.P.Val < 0.05] <- "Downregulated"
rownames(LPC_annotated) <- make.names(LPC_annotated[,"SYMBOL"], unique = TRUE)


LPC_sig <- subset(LPC_annotated, adj.P.Val < 0.05 & abs(logFC)> 1.0) %>% arrange(-logFC)

## create a ranked list of genes for gsea

LPC_gsea <- arrange(LPC_annotated, -logFC)
rownames(LPC_gsea) <- make.names(LPC_gsea[,"SYMBOL"], unique = TRUE)
write.table(LPC_gsea, file="LPC_ranked.csv", sep = ",") # export file for GSEA software


# Venn Diagrams ------

diff_LPC_LSC <- c("LPC", "LSC")               # array with columns to select
diff_LPC_vs_LSC <- diff.exprs[,diff_LPC_LSC]     # create a new class testResult with selected columns

vennDiagram(diff_LPC_vs_LSC, 
            names = c("LPC", "LSC"),
            include = "down",
            circle.col = c("green", "turquoise"))

vennDiagram(diff_LPC_vs_LSC, 
            names = c("LPC", "LSC"),
            include = c("up", "down"),
            circle.col = c("orange", "pink"))



diff_tot <- c("LPC","LSC","HPC")
diff_exprs_tot <- diff.exprs[,diff_tot]     # create a new class testResult with selected columns

vennDiagram(diff_exprs_tot, 
            names = c("LPC","LSC","HPC"),
            include = "down",
            circle.col = c("green", "turquoise", "cyan"))

vennDiagram(diff_exprs_tot, 
            names = c("LPC","LSC","HPC"),
            include = c("up", "down"),
            circle.col = c("orange", "pink", "red"))


# volcano plots ----

make.volcano <- function (a)
{
  ggplot(a, aes(x = logFC, y= -log10(P.Value), color = de)) + 
    scale_color_manual(limits = c("Upregulated", "Downregulated", "Non significant"),
                       values = c("darkred", "darkgreen", "black"), 
                       name = "") +
    xlab(expression(~Log[2]~ "Fold Change"))+
    ylab(expression(~-Log[10]~ "P.Value")) +
    geom_point(size = 2, alpha = 0.5) +
    scale_x_continuous(expand = c(0,0)) +
    scale_y_continuous(expand = c(0,0)) +
    my_theme +
    theme(plot.title = element_text(size = 12, margin = margin(b = 5),hjust=0.5,vjust=0.5, family="Arial", face="bold"))
}

library(ggrepel)

make.volcano(LSC_annotated) +
  labs(title = "LSCs") #+
  geom_text_repel(data = LSC_sig, aes(label = SYMBOL, size = 4, segment.color = "black", segment.alpha = 0.1), 
                  show.legend = FALSE, max.overlaps = Inf, nudge_x = 1.5, nudge_y = - 1, direction = "both", hjust = "right") # label only significant genes

make.volcano(LPC_annotated) +
  labs(title = "LPCs") # +
  geom_text_repel(data = LPC_sig, aes(label = SYMBOL, size = 4, segment.color = "black", segment.alpha = 0.1), 
                show.legend = FALSE, max.overlaps = Inf, nudge_x = 1.5, nudge_y = - 1, direction = "both", hjust = "right") # label only significant genes

  
make.volcano(HPC_annotated) +
  labs(title = "HPCs")
  #geom_text_repel(data = HPC_sig, aes(label = SYMBOL), size = 3, show.legend = FALSE) # label only significant genes  
  

  
# master table
master <- merge(LSC_annotated, LPC_annotated, by = 0)
rownames(master) <- master$Row.names
master <- master[,-1]

master <- merge(master, HPC_annotated, by = 0)
rownames(master) <- master$Row.names
master <- master[,-1]


master <- master[,c(1,2,3, 5,7,8,9,10,14,15,16,17,21)]
names(master) <- c("LogFC.LSC", "P.Value.LSC", "adj.P.Value.LSC", "Symbol", "de.LSC",
                   "LogFC.LPC", "P.Value.LPC", "adj.P.Value.LPC", "de.LPC",
                   "LogFC.HPC", "P.Value.HPC", "adj.P.Value.HPC", "de.HPC")

master <- master[,c(4,1,2,3,5:13)] # reorder columns to have symbols as first


sig_all <- filter(master, 
                  de.LSC == "Upregulated"|
                  de.LSC == "Downregulated"|
                  de.LPC == "Upregulated" |
                  de.LPC == "Downregulated"|
                  de.HPC == "Upregulated"|
                  de.HPC == "Downregulated")

em_annotated <- merge(em_filtered, annotations_final, by = 0)
rownames(em_annotated) <- make.names(em_annotated[,"SYMBOL"], unique = TRUE)
em_annotated <- em_annotated[,- c(1,26,27,28)]

sig_genes <- row.names(sig_all)

em_sig <- em_annotated[sig_genes,]
em_sig <- data.frame(t(scale(t(em_sig))))
data_s <- as.matrix(em_sig)


####### heatmap tot ----

library(ComplexHeatmap) # to make heatmaps
library(circlize) # customize heatmap colors


fontsize <- 0.5  # fontsize for rows (= genes)
col.map <- colorRamp2(c(-2,0,2), c("darkblue", "white", "darkred"))
col.map(seq(-3,3))


# heatmap of all significant genes


Heatmap(data_s, 
        name = "Expression 
  Z-score
        ",
        row_names_side = "left",
        row_names_gp = gpar(cex = fontsize),
        row_dend_side = "left",
        row_dend_width = unit(3, "cm"),
        col = col.map,
        column_split = sample_info$group, 
        cluster_columns = FALSE,
        column_title_gp = gpar(fill = c("limegreen", "darkmagenta", "turquoise4", "royalblue4"), fontsize = 18, color = "black"),
        column_title_side = "bottom",
        column_names_gp = gpar(cex = 0, col = "transparent"),
        column_dend_reorder = TRUE,
        border = TRUE)





# single gene boxplot -------

em.scaled <- data.frame(t(scale(t(em_annotated)))) # use scaled data for representation

make.boxplot <- function(gene)
{
  gene.exp <- em.scaled[gene,]
  gene.exp <- data.frame(t(gene.exp))
  gene.exp$groups = sample_info$group
  names(gene.exp) <- c("gene", "group")
  gene.exp$group <- factor(gene.exp$group, levels = c("HPC", "HSC", "LSC", "LPC"))
  
  ggplot(gene.exp, aes(x = group, y = gene, fill = group, color = group)) + 
    scale_fill_manual(values = c("white", "white", "white", "white")) +
    scale_color_manual(values = c("limegreen", "darkmagenta", "turquoise4", "royalblue4")) +
    geom_boxplot(width = 0.5, size = 0.7) +
    geom_jitter(size = 2) +
    my_theme +
    labs(x = "", y = "Expression (Z-score)")
}


# signle gene targets (no significant since observed on annotated em)

make.boxplot("COX7A2") + labs(title = "COX7A2")
make.boxplot("SHMT2") + labs(title = "SHMT2")


make.boxplot("DDIT4") + labs(title = "DDIT4")
make.boxplot("BNIP3L") + labs(title = "BNIP3L")
make.boxplot("ULK1") + labs(title = "ULK1")


make.boxplot("CD38") + labs(title = "CD38")



# selective gene families

## selected gene lists ----

## slc gene family (LSCs)

slc_gene_family <- filter(LSC_annotated, grepl("SLC", SYMBOL))  # filter rows containing a certain pattern
slc_gene_family.sig <- filter(slc_gene_family, de == "Upregulated" | de == "Downregulated") 

make.volcano(slc_gene_family) + 
  geom_text_repel(data = slc_gene_family.sig, aes(label = SYMBOL), size = 5, show.legend = FALSE) +
  labs(title = "SLC genes (LSCs)") + expand_limits(x = c(0,3), y = c(0,11)) +
  geom_vline(xintercept = c(- 1,1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black")

slc.genes <- slc_gene_family.sig$SYMBOL
slc.sig <- em_annotated[slc.genes,]
slc.sig <- data.frame(t(scale(t(slc.sig))))
slc.sig <- as.matrix(slc.sig)



col.map <- colorRamp2(c(-2,0,2), c("blue", "white", "red"))
col.map(seq(-3,3))


# heatmap of all significant genes


Heatmap(slc.sig, 
        name = "Expression 
  Z-score
        ",
        cluster_columns = FALSE,
        #cluster_rows = FALSE,
        row_names_side = "left",
        row_names_gp = gpar(cex = 1, col = "black"),
        row_dend_side = "left",
        row_dend_width = unit(3, "cm"),
        column_names_gp = gpar(cex = 0, col = "transparent"),
        col = col.map,
        #column_title = "Heatmap tot",
        column_split = sample_info$group, 
        column_title_side = "bottom",
        column_title_gp = gpar(fill = c("limegreen", "darkmagenta", "turquoise4", "royalblue4"), fontsize = 18, col = "black"),
        column_dend_reorder = TRUE,
        border = TRUE)



## slc gene family (LPCs)

slc_gene_family <- filter(LPC_annotated, grepl("SLC", SYMBOL))  # filter rows containing a certain pattern
slc_gene_family.sig <- filter(slc_gene_family, de == "Upregulated" | de == "Downregulated") 

make.volcano(slc_gene_family) + 
  geom_text_repel(data = slc_gene_family.sig, aes(label = SYMBOL), size = 5, show.legend = FALSE, max.overlaps = Inf) +
  labs(title = "SLC genes (LPCs vs LSCs)") + expand_limits(x = c(0,3), y = c(0,11)) +
  geom_vline(xintercept = c(- 1,1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black")



















sessionInfo() # R version 4.0.3

  
  