# Affymetrix microarray differential expression 
# The data analyzed here are a clinical microarray dataset
# of Chronic Myeloid Leukaemia (CML) Leukaemic stem cells (LSCs)
# or Progenitors vs their normal counterparts


# list of packages used

#General Bioconductor packages
library(Biobase)
library(oligoClasses)

#Annotation and data import packages
library(ArrayExpress)
library(pd.hugene.1.0.st.v1)
library(hugene10sttranscriptcluster.db)

#Quality control and pre-processing packages
library(oligo)
library(arrayQualityMetrics)

#Analysis and statistics packages
library(limma)
library(topGO)
library(ReactomePA)
library(clusterProfiler)

#Plotting and color options packages
library(gplots)
library(ggplot2)
library(geneplotter)
library(RColorBrewer)
library(pheatmap)
library(enrichplot)

#Formatting/documentation packages
#library(rmarkdown)
#library(BiocStyle)
library(dplyr)
library(tidyr)

#Helpers:
library(stringr)
library(matrixStats)
library(genefilter)
library(openxlsx)
#library(devtools)

## raw data CEL files download
# the CEL are produced by the array scanner and describe the 
# probe intensities. The data are deposited at ArrayExpress

# anno_AE <- getAE("E-MTAB-2594", type = "raw") # run just once


# the aim is to create an ExpressionSet
# Investigation Description Format (IDF): info about experiment
# Sample and Data Relationship Format (SDRF): info about samples

# import SDRF to get samples info

sdrf_location <- file.path("E-MTAB-2594.sdrf.txt")
SDRF <- read.delim(sdrf_location) # import the SDRF from raw data folder

rownames(SDRF) <- SDRF$Array.Data.File
SDRF <- AnnotatedDataFrame(SDRF) # turn SDRF into AnnotatedDataFrame which will be used to create the ExpressionSet

# create the ExpressionSet
# import the CEL files into the raw_data variable using read.celfiles()
# it is important to import the CEL files in the order that corresponds to
# the SDRF table, thus filenames argument has the $ symbol

raw_data <- read.celfiles(filenames = SDRF$Array.Data.File,
                          verbose = FALSE,
                          phenoData = SDRF)

head(pData(raw_data))
sample_info <- pData(raw_data) # to retrieve sample information
sample_info <- sample_info[,c(1,24,25,28,30)]
colnames(sample_info) <- c("sample_number", "cell_type","compound", "time", "donor")


sample_info$time <- c("7days", "7days", "7days", "7days", "7days", 
                      "7days", "7days", "7days", "7days", "7days", 
                      "7days", "7days", "7days",
                      "8hours", "8hours", "8hours", "8hours", "8hours", 
                      "8hours", "8hours", "8hours", "8hours", "8hours", 
                      "8hours", "8hours", "8hours", "8hours", "8hours", 
                      "8hours", "8hours", "8hours",
                      "none","none", "none", "none", "none",
                      "none", "none", "none", "none", "none",
                      "none", "none", "none", "none", "none",
                      "none", "none", "none")

sample_info$cell_type <- c("CD34+", "CD34+", "CD34+", "CD34+", "CD34+",
                           "CD34+", "CD34+", "CD34+", "CD34+", "CD34+",
                           "CD34+", "CD34+", "CD34+",
                           "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-",
                           "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-",
                           "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-",
                           "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-",
                           "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-",
                           "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-",
                           "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-", "CD34+C38-", 
                           "CD34+C38-")

### Quality control

em <- exprs(raw_data) # retrieve the expression matrix with probes intensities
# log2 transformation
em_log2 <- log2(em)

plotDensities(em, group = sample_info[,"compound"], legend = "topright")
plotDensities(em_log2, group = sample_info[,"compound"], legend = "topright")


# PCA
pca <- prcomp(t(em_log2), scale. = FALSE)
percentVar <- round(100*pca$sdev^2/sum(pca$sdev^2),1)
sd_ratio <- sqrt(percentVar[2]/percentVar[1])

# create data frame for PCA plot
dataGG <- data.frame(PC1 = pca$x[,1],
                     PC2 = pca$x[,2],
                     Compound = sample_info$compound,
                     Cell_type = sample_info$cell_type,
                     Donor = sample_info$donor,
                     Time = as.character(sample_info$time))
library(ggrepel)

ggplot(dataGG, aes(x = PC1, y = PC2, colour = Compound, label = paste("", Donor), shape = Time)) +
  ggtitle("PCA plot of log2-transformed raw data") +
  xlab(paste0("PC1 (", percentVar[1], "%)")) +
  ylab(paste0("PC2 (", percentVar[2], "%)")) +
  geom_point(size = 2) + 
  geom_text_repel() +
  theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 30),
        axis.text.x = element_text(size = 18, colour = "black"),
        axis.title.x = element_text(size =20, colour = "black"),
        axis.text.y = element_text(size = 18, colour = "black"),
        axis.title.y  = element_text(size = 20, colour = "black"),
        panel.border = element_rect(colour = "black", fill = NA, size = 1),
        panel.background = element_blank(),
        legend.key = element_blank(),
        legend.position = "right",
        legend.text = element_text(colour = "black", size = 15),
        legend.title = element_text(colour = "white")
  ) +
  #coord_fixed(ratio = sd_ratio) +
  scale_color_manual(values = c("darkorange2", "dodgerblue4", "darkred", "grey30"))


# the PCA plot shows that both Disease and Cell_Type drive different transcription profiles. the source of variation should be DIsease only

# boxplot of raw_data
# boxplot() from oligo package does the log2-transformation by default, hence the use of raw_data

oligo::boxplot(raw_data, target = "core", main = "Boxplot of log2-intensities for the raw data") # need normalization



# RMA normalization

raw_data_normalized <- rma(raw_data, target = "core")
## Quality assessment of the calibated data
# PCA

em_normalized <- exprs(raw_data_normalized)

pca_normalized <- prcomp(t(em_normalized), scale. = FALSE)
percentVar_normalized <- round(100*pca_normalized$sdev^2/sum(pca_normalized$sdev^2),1)
sd_ratio_normalized <- sqrt(percentVar_normalized[2]/ percentVar_normalized[1])

dataGG_normalized <- data.frame(PC1 = pca_normalized$x[,1], PC2 = pca_normalized$x[,2],
                                Compound = sample_info$compound,
                                Cell_type = sample_info$cell_type,
                                Donor = sample_info$donor,
                                Time = as.character(sample_info$time))

ggplot(dataGG_normalized, aes(x = PC1, y = PC2, colour = Compound, shape = Time)) +
  ggtitle("PCA (normalized)") +
  xlab(paste0("PC1 (", percentVar_normalized[1], "%)")) +
  ylab(paste0("PC2 (", percentVar_normalized[2], "%)")) +
  geom_point(size = 4) +
  theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 30),
        axis.text.x = element_text(size = 18, colour = "black"),
        axis.title.x = element_text(size = 20, colour = "black"),
        axis.text.y = element_text(size = 18, colour = "black"),
        axis.title.y  = element_text(size = 20, colour = "black"),
        panel.border = element_rect(colour = "black", fill = NA, size = 1),
        panel.background = element_blank(),
        legend.key = element_blank(),
        legend.position = "right",
        legend.text = element_text(colour = "black", size = 18),
        legend.title = element_text(colour = "white")
        ) +
  #coord_fixed(ratio = sd_ratio_normalized) +
  scale_color_manual(values = c("darkorange2", "dodgerblue4", "darkred", "grey30"))


oligo::boxplot(raw_data_normalized, target = "core", main = "Boxplot of log2-intensities for the normalized raw data") # data are normalized now
 
library(ggrepel)
library(sva)

# correction for batch effect with ComBat

em_normalized.m <- as.matrix(em_normalized)
combat_model <- model.matrix(~ 1, sample_info)
em_corrected <- ComBat(dat = em_normalized.m, batch = sample_info$donor, mod = combat_model, par.prior = TRUE, prior.plots = FALSE)

pca_corrected <- prcomp(t(em_corrected), scale. = FALSE)

percentVar_corrected <- round(100*pca_corrected$sdev^2/sum(pca_corrected$sdev^2),1)
sd_ratio_corrected <- sqrt(percentVar_corrected[2]/ percentVar_corrected[1])

dataGG_corrected <- data.frame(PC1 = pca_corrected$x[,1], PC2 = pca_corrected$x[,2],
                               Compound = sample_info$compound,
                               Cell_type = sample_info$cell_type,
                               Donor = sample_info$donor,
                               Time = as.character(sample_info$time))


# PCA plot (FINAL) -----

ggplot(dataGG_corrected, aes(x = PC1, y = PC2, colour = Compound, shape = Time)) +
  ggtitle("PCA (normalized and batch-corrected)") +
  xlab(paste0("PC1 (", percentVar_corrected[1], "%)")) +
  ylab(paste0("PC2 (", percentVar_corrected[2], "%)")) +
  geom_point(size = 4) +
  #geom_text_repel(size = 4, max.overlaps = Inf) +
  scale_shape_manual(values=c(16, 17, 3)) +
  theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 30),
        axis.text.x = element_text(size = 18, colour = "black"),
        axis.title.x = element_text(size = 20, colour = "black"),
        axis.text.y = element_text(size = 18, colour = "black"),
        axis.title.y  = element_text(size = 20, colour = "black"),
        panel.border = element_rect(colour = "black", fill = NA, size = 1),
        panel.background = element_blank(),
        legend.key = element_blank(),
        legend.position = "right",
        legend.text = element_text(colour = "black", size = 18),
        legend.title = element_text(colour = "white")
  ) +
  #coord_fixed(ratio = sd_ratio_normalized) +
  scale_color_manual(values = c("darkorange2", "dodgerblue4", "darkred", "grey30"))



# filter lowly expressed genes

medians <- rowMedians(em_corrected)

# histogram of median intensities
hist(em_corrected, 100, col = "cornsilk1", freq = FALSE,
     main = "Histogram of the median intensities",
     border = "antiquewhite4",
     xlab = "Median intensities")

# alternative way of plotting the intensities
plotDensities(em_corrected, group = sample_info[,"compound"],legend = "topright")

cut_off <- 5                  # cut-off for the peak
abline(v = cut_off, col = "coral4", lwd = 2)

# transcripts without intensities larger than the threshold in at least as many arrays as the smallest experimental group are excluded

no_of_samples <- 
  table(paste0(pData(raw_data_normalized)$Factor.Value.compound., "", 
               pData(raw_data_normalized)$Factor.Value.cell.type.))
no_of_samples 
samples_cutoff <- min(no_of_samples) # min number of replicates per sample group

idx_threshold <- apply(em_corrected, 1, 
                       function(x){
                         sum(x > cut_off) >= samples_cutoff}) # applies the idx_threshold function to each row and excludes the ones that are not in at least 6 replicates

table(idx_threshold)
raw_data_filtered <- subset(raw_data_normalized,idx_threshold)
em_filtered <- subset(em_corrected, idx_threshold)
plotDensities(em_filtered, group = sample_info[,"compound"],legend = "topright")


## Retrieve annotations
# add the feauture data (i.e. annotations) to the filtered ExpressionSet

annotation_raw_data <- AnnotationDbi::select(hugene10sttranscriptcluster.db,
                                             keys = (featureNames(raw_data_filtered)),
                                             columns = c("SYMBOL", "GENENAME"),
                                             keytype = "PROBEID")

# select() from Annotation Dbi to retrieve gene symbols and a short description for each probe

annotations <- subset(annotation_raw_data, !is.na(SYMBOL)) # get rid of all the NAs for SYMBOLs

## remove multiple mappings
# many gens will map to multiple gene symbols

anno_grouped <- group_by(annotations, PROBEID)
anno_summarized <- summarize(anno_grouped, no_of_matches = n_distinct(SYMBOL))

head(anno_summarized)
anno_filtered <- filter(anno_summarized, no_of_matches >1)
head(anno_filtered)
probe_stats <- anno_filtered 

nrow(probe_stats)

ids_to_exlude <- (featureNames(raw_data_filtered) %in% probe_stats$PROBEID)
table(ids_to_exlude)

raw_data_final <- subset(raw_data_filtered, !ids_to_exlude)
validObject(raw_data_final)

head(annotations)

fData(raw_data_final)$PROBEID <- rownames(fData(raw_data_final))
fData(raw_data_final) <- left_join(fData(raw_data_final), annotations)

rownames(fData(raw_data_final)) <- fData(raw_data_final)$PROBEID 
validObject(raw_data_final)

annotations_final <- fData(raw_data_final)
annotations_final <- na.omit(annotations_final)


# differentially expressed genes


## limma pipeline

group <- with(sample_info,                                             # with() to refer to the columns in samples dataframe
              paste(compound,time, sep = "."))  

design <- model.matrix(~0 + group)

colnames(design) <- c("dasatinib.7d", "dasatinib.8h",
                      "imatinib.7d", "imatinib.8h",
                      "nilotinib.7d", "nilotinib.8h",
                      "ctrl")

fit <- lmFit(em_filtered, design)  # fit linear model to samples

contrasts <- makeContrasts(dasatinib.7d = "dasatinib.7d - ctrl",
                           imatinib.7d = "imatinib.7d - ctrl",
                           nilotinib.7d = "nilotinib.7d - ctrl",
                           dasatinib.8h = "dasatinib.8h - ctrl",
                           imatinib.8h = "imatinib.8h - ctrl",
                           nilotinib.8h = "nilotinib.8h - ctrl",
                           dasatinib = "(dasatinib.7d - ctrl) + ( dasatinib.8h - ctrl)",
                           imatinib = "(imatinib.7d - ctrl) + (imatinib.8h - ctrl)",
                           nilotinib = "(nilotinib.7d - ctrl) + (nilotinib.8h - ctrl)",
                           levels = design)

fit2 <- contrasts.fit(fit, contrasts)
fit2 <- eBayes(fit2)

diff.exprs <- decideTests(fit2, p.value = 0.05, adjust.method = "fdr", lfc = 1) 
summary(diff.exprs)  # list of up/down regulated and non significant genes


# extract results

# imatinib ----

imatinib <- topTable(fit2, coef = "imatinib.7d", number = Inf) # retrieve a data.frame of differential expression

hist(imatinib$P.Value, col = brewer.pal(3, name = "Set2")[2], main = "imatinib", xlab = "p-values")

imatinib_annotated <- merge(imatinib, annotations_final, by = 0)
row.names(imatinib_annotated) <- imatinib_annotated$Row.names
imatinib_annotated <- imatinib_annotated[, c(2,5,6,8,9,10)]

imatinib_annotated$de <- "Non significant"
imatinib_annotated$de[imatinib_annotated$logFC > 1 & imatinib_annotated$adj.P.Val < 0.05] <- "Upregulated"
imatinib_annotated$de[imatinib_annotated$logFC < -1 & imatinib_annotated$adj.P.Val < 0.05] <- "Downregulated"

imatinib_sig <- subset(imatinib_annotated, adj.P.Val < 0.05 & abs(logFC)> 1.0) %>% arrange(-logFC)

## create a ranked list of genes for gsea

imatinib_gsea <- arrange(imatinib_annotated, -logFC)
rownames(imatinib_gsea) <- make.names(imatinib_gsea[,"SYMBOL"], unique = TRUE)

imatinib_gsea$pi_score <- imatinib_gsea$logFC * -log10(imatinib_gsea$P.Value)
imatinib_gsea <- arrange(imatinib_gsea, -pi_score)
write.table(imatinib_gsea, file="imatinib_ranked.csv", sep = ",") # export file for GSEA software


# dasatinib ----

dasatinib <- topTable(fit2, coef = "dasatinib.7d", number = Inf) # retrieve a data.frame of differential expression

hist(dasatinib$P.Value, col = brewer.pal(3, name = "Set2")[1], main = "dasatinib", xlab = "p-values")

dasatinib_annotated <- merge(dasatinib, annotations_final, by = 0)
row.names(dasatinib_annotated) <- dasatinib_annotated$Row.names
dasatinib_annotated <- dasatinib_annotated[, c(2,5,6,8,9,10)]

dasatinib_annotated$de <- "Non significant"
dasatinib_annotated$de[dasatinib_annotated$logFC > 1 & dasatinib_annotated$adj.P.Val < 0.05] <- "Upregulated"
dasatinib_annotated$de[dasatinib_annotated$logFC < -1 & dasatinib_annotated$adj.P.Val < 0.05] <- "Downregulated"

dasatinib_sig <- subset(dasatinib_annotated, adj.P.Val < 0.05 & abs(logFC)> 1.0) %>% arrange(-logFC)

## create a ranked list of genes for gsea

dasatinib_gsea <- arrange(dasatinib_annotated, -logFC)
rownames(dasatinib_gsea) <- make.names(dasatinib_gsea[,"SYMBOL"], unique = TRUE)

dasatinib_gsea$pi_score <- dasatinib_gsea$logFC * -log10(dasatinib_gsea$P.Value)
dasatinib_gsea <- arrange(dasatinib_gsea, -pi_score)
write.table(dasatinib_gsea, file="dasatinib_ranked.csv", sep = ",") # export file for GSEA software

# nilotinib ----

nilotinib <- topTable(fit2, coef = "nilotinib.7d", number = Inf) # retrieve a data.frame of differential expression

hist(nilotinib$P.Value, col = brewer.pal(3, name = "Set2")[3], main = "nilotinib", xlab = "p-values")

nilotinib_annotated <- merge(nilotinib, annotations_final, by = 0)
row.names(nilotinib_annotated) <- nilotinib_annotated$Row.names
nilotinib_annotated <- nilotinib_annotated[, c(2,5,6,8,9,10)]

nilotinib_annotated$de <- "Non significant"
nilotinib_annotated$de[nilotinib_annotated$logFC > 1 & nilotinib_annotated$adj.P.Val < 0.05] <- "Upregulated"
nilotinib_annotated$de[nilotinib_annotated$logFC < -1 & nilotinib_annotated$adj.P.Val < 0.05] <- "Downregulated"

nilotinib_sig <- subset(nilotinib_annotated, adj.P.Val < 0.05 & abs(logFC)> 1.0) %>% arrange(-logFC)

## create a ranked list of genes for gsea

nilotinib_gsea <- arrange(nilotinib_annotated, -logFC)
rownames(nilotinib_gsea) <- make.names(nilotinib_gsea[,"SYMBOL"], unique = TRUE)
nilotinib_gsea$pi_score <- nilotinib_gsea$logFC * -log10(nilotinib_gsea$P.Value)
nilotinib_gsea <- arrange(nilotinib_gsea, -pi_score)
write.table(nilotinib_gsea, file="nilotinib_ranked.csv", sep = ",") # export file for GSEA software


# volcano plots ----

make.volcano <- function (a)
{
  ggplot(a, aes(x = logFC, y= -log10(P.Value), col= de)) + 
    scale_color_manual(limits = c("Upregulated", "Downregulated", "Non significant"),
                       values = c("red", "blue", "gray"), 
                       name = "") +
    xlab(expression(~Log[2]~ "Fold Change"))+
    ylab(expression(~-Log[10]~ "P.Value")) +
    geom_point(size = 2) +
    scale_x_continuous(expand = c(0,0)) +
    scale_y_continuous(expand = c(0,0)) +
    theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 30),
          axis.text.x = element_text(size = 18, colour = "black"),
          axis.title.x = element_text(size = 20, colour = "black"),
          axis.text.y = element_text(size = 18, colour = "black"),
          axis.title.y  = element_text(size = 20, colour = "black"),
          axis.line = element_line(colour = "black", size = 1),
          panel.background = element_blank(),
          legend.key = element_blank(),
          legend.position = "bottom",
          legend.text = element_text(colour = "black", size = 15))
}

library(ggrepel)

make.volcano(dasatinib_annotated) +
  labs(title = "dasatinib vs ctrl")# +
  geom_text_repel(data = dasatinib_sig, aes(label = SYMBOL), size = 3, show.legend = FALSE, max.overlaps = Inf) # label only significant genes

make.volcano(imatinib_annotated) +
  labs(title = "imatinib vs ctrl")# + 
  geom_text_repel(data = imtinib_sig, aes(label = SYMBOL), size = 3, show.legend = FALSE, max.overlaps = Inf) # label only significant genes

make.volcano(nilotinib_annotated) +
  labs(title = "nilotinib vs ctrl")# + 
  geom_text_repel(data = nilotinib_sig, aes(label = SYMBOL), size = 3, show.legend = FALSE, max.overlaps = Inf) # label only significant genes  


# master table
master <- merge(dasatinib_annotated, imatinib_annotated, by = 0)
rownames(master) <- master$Row.names
master <- master[,-1]

master <- merge(master, nilotinib_annotated, by = 0)
rownames(master) <- master$Row.names
master <- master[,-1]

rownames(master) <- make.names(master[,"SYMBOL.x"], unique = TRUE)
master <- master[,c(1,2,3, 5,7,8,9,10,14,15,16,17,21)]
names(master) <- c("LogFC.dasatinib", "P.Value.dasatinib", "adj.P.Value.dasatinib", "Symbol", "de.dasatinib",
                   "LogFC.imatinib", "P.Value.imatinib", "adj.P.Value.imatinib", "de.imatinib",
                   "LogFC.nilotinib", "P.Value.nilotinib", "adj.P.Value.nilotinib", "de.nilotinib")

master <- master[,c(4,1,2,3,5:13)] # reorder columns to have symbols as first



write.table(master, file ="master_table_LSC_TKIs.csv", sep = ",", 
            row.names = FALSE, col.names = TRUE)


sig_all <- filter(master, 
                  de.dasatinib == "Upregulated"|
                  de.dasatinib == "Downregulated"|
                  de.imatinib == "Upregulated" |
                  de.imatinib == "Downregulated"|
                  de.nilotinib == "Upregulated"|
                  de.nilotinib == "Downregulated")

em_annotated <- merge(em_filtered, annotations_final, by = 0)
rownames(em_annotated) <- make.names(em_annotated[,"SYMBOL"], unique = TRUE)
em_annotated <- em_annotated[,- c(1,51,52,53)]

sig_genes <- row.names(sig_all)

em_sig <- em_annotated[sig_genes,]
em_sig <- data.frame(t(scale(t(em_sig))))
data_s <- as.matrix(em_sig)


####### heatmap tot ----

library(ComplexHeatmap) # to make heatmaps
library(circlize) # customize heatmap colors


fontsize <- 0.5  # fontsize for rows (= genes)
col.map <- colorRamp2(c(-2,0,2), c("blue", "black", "red"))
col.map(seq(-3,3))


# heatmap of all significant genes

sample_info$compound <- factor(sample_info$compound, levels = c("none", "imatinib", "dasatinib", "nilotinib")) # to set up the desired order

Heatmap(data_s, 
        name = "Expression 
  Z-score
        ",
        row_names_side = "left",
        row_names_gp = gpar(cex = fontsize),
        row_dend_side = "left",
        row_dend_width = unit(3, "cm"),
        col = col.map,
        column_split = sample_info$compound, 
        column_title_gp = gpar(fill = c("grey30", "dodgerblue4", "darkorange2", "darkred"), fontsize = 18, color = "black"),
        column_title_side = "bottom",
        column_names_gp = gpar(cex = 0.2, col = "transparent"),
        cluster_columns = FALSE,
        column_dend_reorder = TRUE,
        border = TRUE)

################### vennDiagrams ----

diff_TKIs <- c("dasatinib", "imatinib", "nilotinib")    # array with columns to select
diff_TKIs <- diff.exprs[,diff_TKIs]     # create a new class testResult with selected columns

vennDiagram(diff_TKIs, 
            names = c("dasatinib", "imatinib", "nilotinib"),
            include = c("up","down"),
            circle.col = c("green", "turquoise", "blue"))

#######################################################################################################################

### pathway enrichment analysis (imatinb)

rownames(imatinib_sig) <- make.names(imatinib_sig[,"SYMBOL"], unique = TRUE) # row.names as SYMBOL

# convert SYMBOL to ENTREZ
entrez_ids_imatinib <- mapIds(hugene10sttranscriptcluster.db, 
                               keys = rownames(imatinib_sig), 
                               keytype = "SYMBOL",
                               column = "ENTREZID")



# Reactome enrichment
reactome_enrich.imatinib <- enrichPathway(gene = entrez_ids_imatinib,
                                           readable = TRUE,
                                           organism = "human",
                                           pvalueCutoff = 0.05,
                                           qvalueCutoff = 0.9)

gene.ID.reactome.imatinib <- reactome_enrich.imatinib$geneID           # extract gene sets
description.reactome.imatinib <- reactome_enrich.imatinib$Description    # extract description
p.adj.reactom.imatinib <- reactome_enrich.imatinib$p.adjust             # extract p.adjust

# create a data frame for top genes
reactome_results_imatinib <- data.frame(cbind(gene.ID.reactome.imatinib, p.adj.reactom.imatinib))
reactome_results_imatinib$enriched.gene.number <- reactome_enrich.imatinib$Count
reactome_results_imatinib$pathway.name <- description.reactome.imatinib

write.table(reactome_results_imatinib, file ="reactome_results_imatinib.csv", sep = ",", 
            row.names = FALSE, col.names = TRUE)

# ORA
go_enrich.imatinib <- enrichGO(gene = entrez_ids_imatinib,
                                OrgDb = org.Hs.eg.db,
                                readable = TRUE,
                                ont = "BP",
                                pvalueCutoff = 0.05,
                                qvalueCutoff = 0.10)

gene.sets.imatinib <- go_enrich.imatinib$geneID           # extract gene sets
description.imatinib <- go_enrich.imatinib$Description    # extract description
p.adj.imatinib <- go_enrich.imatinib$p.adjust             # extract p.adjust

# create a data frame for top genes
ora_results_imatinib <- data.frame(cbind(gene.sets.imatinib, p.adj.imatinib))
ora_results_imatinib$enriched.gene.number <- go_enrich.imatinib$Count
ora_results_imatinib$pathway.name <- description.imatinib

write.table(ora_results_imatinib, file ="ora_results_imatinib.csv", sep = ",", 
            row.names = FALSE, col.names = TRUE)

### pathway enrichment analysis (dasatinib)

rownames(dasatinib_sig) <- make.names(dasatinib_sig[,"SYMBOL"], unique = TRUE) # row.names as SYMBOL

# convert SYMBOL to ENTREZ
entrez_ids_dasatinib <- mapIds(hugene10sttranscriptcluster.db, 
                              keys = rownames(dasatinib_sig), 
                              keytype = "SYMBOL",
                              column = "ENTREZID")



# Reactome enrichment
reactome_enrich.dasatinib <- enrichPathway(gene = entrez_ids_dasatinib,
                                          readable = TRUE,
                                          organism = "human",
                                          pvalueCutoff = 0.05,
                                          qvalueCutoff = 0.9)

gene.ID.reactome.dasatinib <- reactome_enrich.dasatinib$geneID           # extract gene sets
description.reactome.dasatinib <- reactome_enrich.dasatinib$Description    # extract description
p.adj.reactom.dasatinib <- reactome_enrich.dasatinib$p.adjust             # extract p.adjust

# create a data frame for top genes
reactome_results_dasatinib <- data.frame(cbind(gene.ID.reactome.dasatinib, p.adj.reactom.dasatinib))
reactome_results_dasatinib$enriched.gene.number <- reactome_enrich.dasatinib$Count
reactome_results_dasatinib$pathway.name <- description.reactome.dasatinib

write.table(reactome_results_dasatinib, file ="reactome_results_dasatinib.csv", sep = ",", 
            row.names = FALSE, col.names = TRUE)

# ORA
go_enrich.dasatinib <- enrichGO(gene = entrez_ids_dasatinib,
                               OrgDb = org.Hs.eg.db,
                               readable = TRUE,
                               ont = "BP",
                               pvalueCutoff = 0.05,
                               qvalueCutoff = 0.10)

gene.sets.dasatinib <- go_enrich.dasatinib$geneID           # extract gene sets
description.dasatinib <- go_enrich.dasatinib$Description    # extract description
p.adj.dasatinib <- go_enrich.dasatinib$p.adjust             # extract p.adjust

# create a data frame for top genes
ora_results_dasatinib <- data.frame(cbind(gene.sets.dasatinib, p.adj.dasatinib))
ora_results_dasatinib$enriched.gene.number <- go_enrich.dasatinib$Count
ora_results_dasatinib$pathway.name <- description.dasatinib

write.table(ora_results_dasatinib, file ="ora_results_dasatinib.csv", sep = ",", 
            row.names = FALSE, col.names = TRUE)



### pathway enrichment analysis (nilotinib)

rownames(nilotinib_sig) <- make.names(nilotinib_sig[,"SYMBOL"], unique = TRUE) # row.names as SYMBOL

# convert SYMBOL to ENTREZ
entrez_ids_nilotinib <- mapIds(hugene10sttranscriptcluster.db, 
                     keys = rownames(nilotinib_sig), 
                     keytype = "SYMBOL",
                     column = "ENTREZID")



# Reactome enrichment
reactome_enrich.nilotinib <- enrichPathway(gene = entrez_ids_nilotinib,
                                    readable = TRUE,
                                    organism = "human",
                                    pvalueCutoff = 0.05,
                                    qvalueCutoff = 0.9)




gene.ID.reactome.nilotinib <- reactome_enrich.nilotinib$geneID           # extract gene sets
description.reactome.nilotinib <- reactome_enrich.nilotinib$Description    # extract description
p.adj.reactom.nilotinib <- reactome_enrich.nilotinib$p.adjust             # extract p.adjust

# create a data frame for top genes
reactome_results_nilotinib <- data.frame(cbind(gene.ID.reactome.nilotinib, p.adj.reactom.nilotinib))
reactome_results_nilotinib$enriched.gene.number <- reactome_enrich.nilotinib$Count
reactome_results_nilotinib$pathway.name <- description.reactome.nilotinib


# ORA
go_enrich.nilotinib <- enrichGO(gene = entrez_ids_nilotinib,
                                    OrgDb = org.Hs.eg.db,
                                    readable = TRUE,
                                    ont = "BP",
                                    pvalueCutoff = 0.05,
                                    qvalueCutoff = 0.10)

gene.sets.nilotinib <- go_enrich.nilotinib$geneID           # extract gene sets
description.nilotinib <- go_enrich.nilotinib$Description    # extract description
p.adj.nilotinib <- go_enrich.nilotinib$p.adjust             # extract p.adjust

# create a data frame for top genes
ora_results_nilotinib <- data.frame(cbind(gene.sets.nilotinib, p.adj.nilotinib))
ora_results_nilotinib$enriched.gene.number <- go_enrich.nilotinib$Count
ora_results_nilotinib$pathway.name <- description.nilotinib

write.table(ora_results_nilotinib, file ="ora_results_nilotinib.csv", sep = ",", 
            row.names = FALSE, col.names = TRUE)

####################################### GSEA -------  

library(fgsea) # gsea analysis
library(data.table)
#library(GSEABase)
library(msigdbr) # loading pathways

# loading pathways

# load pathways hallmarks gene set

h.gene.sets <- msigdbr(species = "Homo sapiens", category ="H")
head(h.gene.sets)

pathwaysH <- split(x = h.gene.sets$gene_symbol, f = h.gene.sets$gs_name)

# load pathway c2 gene set

c2.gene.sets <- msigdbr(species = "Homo sapiens", category ="C2")
head(c2.gene.sets)

pathwaysC2 <- split(x = c2.gene.sets$gene_symbol, f = c2.gene.sets$gs_name)

# load pathways c5 gene set

c5.gene.sets = msigdbr(species = "Homo sapiens", category ="C5")
head(c5.gene.sets)

pathwaysC5 <- split(x = c5.gene.sets$gene_symbol, f = c5.gene.sets$gs_name)



gsea_barplot <- function (x)
{
  ggplot(x, aes(x = logFC, y = SYMBOL, fill = de)) + 
    geom_col() +
    theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 30),
          axis.text.x = element_text(size = 18, colour = "black"),
          axis.title.x = element_text(size = 20, colour = "black"),
          axis.text.y = element_text(size = 18, colour = "black"),
          axis.title.y  = element_text(size = 20, colour = "black"),
          axis.line = element_line(colour = "black", size = 1),
          panel.background = element_blank(),
          legend.key = element_blank(),
          legend.position = "bottom",
          legend.text = element_text(size = 16, colour = "black"),
          legend.title = element_blank(),
          plot.caption = element_text(size = 16, colour = "black", hjust = 0.5)
    ) +
    xlab(expression(~Log[2]~ "Fold Change"))+
    ylab("") +
    scale_fill_manual(limits = c("Upregulated", "Downregulated"), values = c("darkred", "darkgreen")) +
    geom_vline(xintercept = 0, size = 1, color = "black") 
}

## GSEA (imatinib)

dt_imatinib <- data.table(imatinib_annotated, keep.rownames = T)

ranks_imatinib <- dt_imatinib$logFC 
names(ranks_imatinib) <- dt_imatinib$SYMBOL
head(ranks_imatinib, 20)
barplot(sort(ranks_imainib, decreasing = TRUE))     # plot the ranked logFC

# GSEA analysis pathwaysH

fgsea_imatinib.H <- fgsea(pathwaysH, 
                     ranks_imatinib, 
                     minSize=15, 
                     maxSize = 500,
                     nperm= 1000)

head(fgsea_imatinib.H[order(padj, -abs(NES)), ], n=10) # look at the top 10 results

# enrichment score plots (select pathway of interest)

plotEnrichment(pathwaysH[["HALLMARK_ANGIOGENESIS"]], ranks_imatinib) + labs(title = "HALLMARK_ANGIOGENESIS", 
                                                                                    caption = "NES = 1.8298431
                                                                                           adj.P.value = 0.013185654

                                                                                           ",
                                                                                    x = "",
                                                                                    y = "Enrichment Score") +
  theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 30),
        axis.text.x = element_text(size = 18, colour = "black"),
        axis.title.x = element_text(size = 20, colour = "black"),
        axis.text.y = element_text(size = 18, colour = "black"),
        axis.title.y  = element_text(size = 18, colour = "black"),
        axis.line = element_line(colour = "black", size = 1),
        panel.background = element_blank(),
        plot.caption = element_text(size = 16, colour = "black")
  )



gsea.imatinib <- fgsea_imatinib.H[36,8]
gsea.imatinib <- as.character(unlist(gsea.imatinib, ","))

rownames(imatinib_sig) <- make.names(imatinib_sig[,"SYMBOL"], unique = TRUE) # this way due to duplicate row names 

imatinib_sig.h <- unique(imatinib_sig[gsea.imatinib,]) %>% na.omit(imatinib_sig.h)
gsea_barplot(imatinib_sig.h) + labs(title = "HALLMARK_OXIDATIVE_PHOSPHORYLATION") 





### GSEA dasatinib 

dt_dasatinib <- data.table(dasatinib_annotated, keep.rownames = T)

ranks_dasatinib <- dt_dasatinib$logFC 
names(ranks_dasatinib) <- dt_dasatinib$SYMBOL
head(ranks_dasatinib, 20)
barplot(sort(ranks_dasatinib, decreasing = TRUE))     # plot the ranked logFC

# GSEA analysis pathwaysH

fgsea_dasatinib.H <- fgsea(pathwaysH, 
                          ranks_dasatinib, 
                          minSize=15, 
                          maxSize = 500,
                          nperm= 1000)

head(fgsea_dasatinib.H[order(padj, -abs(NES)), ], n=10) # look at the top 10 results

# enrichment score plots (select pathway of interest)

plotEnrichment(pathwaysH[[""]], ranks_dasatinib) + labs(title = "", 
                                                       caption = "NES = 
                                                                                           P.value = 

                                                                                           ",
                                                       x = "",
                                                       y = "Enrichment Score") +
  theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 30),
        axis.text.x = element_text(size = 18, colour = "black"),
        axis.title.x = element_text(size = 20, colour = "black"),
        axis.text.y = element_text(size = 18, colour = "black"),
        axis.title.y  = element_text(size = 18, colour = "black"),
        axis.line = element_line(colour = "black", size = 1),
        panel.background = element_blank(),
        plot.caption = element_text(size = 16, colour = "black")
  )



gsea.dasatinib <- fgsea_dasatinib.H[36,8]
gsea.dasatinib <- as.character(unlist(gsea.dasatinib, ","))

rownames(dasatinib_sig) <- make.names(dasatinib_sig[,"SYMBOL"], unique = TRUE) # this way due to duplicate row names 

dasatinib_sig.h <- unique(dasatinib_sig[gsea.dasatinib,]) %>% na.omit(dasatinib_sig.h)
gsea_barplot(dasatinib_sig.h) + labs(title = "HALLMARK_OXIDATIVE_PHOSPHORYLATION") 



### GSEA nilotinib

dt_nilotinib <- data.table(nilotinib_annotated, keep.rownames = T)

ranks_nilotinib <- dt_nilotinib$logFC 
names(ranks_nilotinib) <- dt_nilotinib$SYMBOL
head(ranks_nilotinib, 20)
barplot(sort(ranks_nilotinib, decreasing = TRUE))     # plot the ranked logFC

# GSEA analysis pathwaysH

fgsea_nilotinib.H <- fgsea(pathwaysH, 
                           ranks_nilotinib, 
                           minSize=15, 
                           maxSize = 500,
                           nperm= 1000)

head(fgsea_nilotinib.H[order(padj, -abs(NES)), ], n=10) # look at the top 10 results

# enrichment score plots (select pathway of interest)

plotEnrichment(pathwaysH[[""]], ranks_nilotinib) + labs(title = "", 
                                                        caption = "NES = 
                                                                                           P.value = 

                                                                                           ",
                                                        x = "",
                                                        y = "Enrichment Score") +
  theme(plot.title = element_text(hjust = 0.5, colour = "black", size = 30),
        axis.text.x = element_text(size = 18, colour = "black"),
        axis.title.x = element_text(size = 20, colour = "black"),
        axis.text.y = element_text(size = 18, colour = "black"),
        axis.title.y  = element_text(size = 18, colour = "black"),
        axis.line = element_line(colour = "black", size = 1),
        panel.background = element_blank(),
        plot.caption = element_text(size = 16, colour = "black")
  )



gsea.nilotinib <- fgsea_nilotinib.H[7,8]
gsea.nilotinib <- as.character(unlist(gsea.nilotinib, ","))

rownames(nlotinib_sig) <- make.names(nilotinib_sig[,"SYMBOL"], unique = TRUE) # this way due to duplicate row names 

nilotinib_sig.h <- unique(nilotinib_sig[gsea.nilotinib,]) %>% na.omit(nilotinib_sig.h)
gsea_barplot(nilotinib_sig.h) + labs(title = "HALLMARK_APOPTOSIS") 

sessionInfo() # R version 4.0.3