---
title: Pyruvate Anaplerosis is a Targetable Vulnerability in Persistent Leukaemic
  Stem Cells
Analyst: Ryan SY Kwan
date: "Update 2023.06.21"
output:
  word_document:
    toc: yes
    toc_depth: '4'
  html_notebook:
    toc: yes
    toc_depth: '4'
    number_sections: yes
    toc_float:
      collapsed: no
  html_document:
    toc: yes
    toc_depth: '4'
    df_print: paged
    number_sections: yes
    toc_float:
      collapsed: no
subtitle: 2023.02.25 Multivariable survival analysis for AML
---

# Introduction

PI Name: Vignir Helgason

Main Experimental Contact: Kevin Rattigan

**Aim**

Does the gene PC have any bearing on survival in adult (TGCA) or paediatric (Target) AML? This will involve multivariate analysis factoring gender, subtype, cytogenetic risk categories and treatment.

The study's objective is to examine the potential impact of the PC gene on survival in adult and pediatric acute myeloid leukemia (AML) using multivariate analysis. Various factors, including gender, AML subtype, cytogenetic risk categories, and treatment, are taken into consideration. The feedback received suggests a strong association between PC expression and survival in pediatric AML, while no significant association is found in adult AML. Notably, a high combined expression of MPC1 and MPC2 is linked to poor survival in adult AML. Concerns are raised regarding the multivariate analysis and potential confounding factors like treatment protocols and patient inclusion/exclusion criteria. To address these concerns, Kevin suggests analyzing the survival curves of PC, MPC1, and MPC2 in both pediatric and adult AML, while considering gender, cytogenetic risk categories, and treatment. The study may exclude specific subtypes, such as APML, t(15;17), and chronic myeloid leukemia (CML) with BCR-ABL1, while emphasizing the importance of FLT3-ITD status and cytogenetic risk categories (poor, intermediate, and good).

# Load data

```{r}
suppressMessages({
  library(dplyr)
  library(readxl)
  library(survival)
  library(survminer)
  library(ggplot2)
  library(cowplot)
  library(ComplexHeatmap)
})
```

## TCGA-LAML

### Clinical phenotype

#### UCSC Xena Browswer

[GDC TCGA Acute Myeloid Leukemia (LAML)](https://xenabrowser.net/datapages/?dataset=TCGA-LAML.GDC_phenotype.tsv&host=https%3A%2F%2Fgdc.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)

```{r}
clinical_TCGA <- read.csv("../Xena browser/TCGA-LAML.GDC_phenotype.tsv", sep="\t")

head(clinical_TCGA)
```

#### cBioPortal

[Acute Myeloid Leukemia (TCGA, NEJM 2013)](https://www.cbioportal.org/study?id=laml_tcga_pub)

```{r}
clinical_TCGA2 = read.csv("../cBioPortal/laml_tcga_pub_clinical_data.tsv", sep="\t")

head(clinical_TCGA2)
```


### Gene expression

[HTSeq-FPKM-UQ from UCSC Xena Browser](https://xenabrowser.net/datapages/?dataset=TCGA-LAML.htseq_fpkm-uq.tsv&host=https%3A%2F%2Fgdc.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)

The upper quartile FPKM (FPKM-UQ) is a modified FPKM calculation in which the protein coding gene in the 75th percentile position is substituted for the sequencing quantity. This is thought to provide a more stable value than including the noisier genes at the extremes. [link](https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/)

```{r}
gene_TCGA <- read.csv("../Xena browser/TCGA-LAML.htseq_fpkm-uq.tsv", 
                      sep="\t", check.names = FALSE, row.names = 1)
rownames(gene_TCGA) <- sub("\\..*$", "", rownames(gene_TCGA))
gene_TCGA[1:3,1:3]

# Keep the 151 samples with RNA-seq from clinical data
table(clinical_TCGA$submitter_id.samples %in% colnames(gene_TCGA))
table(clinical_TCGA2$Sample.ID %in% sub(".$","",colnames(gene_TCGA)))

clinical_TCGA <- clinical_TCGA[match(colnames(gene_TCGA), clinical_TCGA$submitter_id.samples),]
clinical_TCGA2 <- clinical_TCGA2[match(sub(".$","",colnames(gene_TCGA)),
                                       clinical_TCGA2$Sample.ID),]
```

### 29 patients with FLT3-ITD positive

Figure 8A of [A clinical transcriptome approach to patient stratification and therapy selection in acute myeloid leukemia](https://doi.org/10.1038/s41467-021-22625-y).

```{r}
df_tem <- read_excel("../TCGA_T. Roderick Docking/41467_2021_22625_MOESM6_ESM.xlsx", 
                     sheet = "Figure 8A - TCGA")
head(df_tem)

clinical_TCGA$FLT3_ITD <- factor(clinical_TCGA$submitter_id %in% df_tem$tcga_id[df_tem$FLT3_ITD=="positive"], 
                                 levels = c(FALSE, TRUE))

table(clinical_TCGA$FLT3_ITD)
```

## TARGET-AML

### Clinical phenotype

#### UCSC Xena Browswer

[UCSC Xena Browser](https://xenabrowser.net/datapages/?dataset=TARGET-AML.clinical.tsv&host=https%3A%2F%2Fgdc.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)

```{r}
clinical_TARGET <- read.csv("../Xena browser/TARGET-AML.clinical.tsv", sep="\t")

head(clinical_TARGET)
```

#### cBioPortal

[Acute Myeloid Leukemia (TCGA, NEJM 2013)](https://www.cbioportal.org/study?id=laml_tcga_pub)

```{r}
clinical_TARGET2 = read.csv("../cBioPortal/aml_target_2018_pub_clinical_data.tsv", sep="\t")

head(clinical_TARGET2)

table(clinical_TARGET$TARGET.USI %in% clinical_TARGET2$Patient.ID)
```

### Gene expression

[HTSeq-FPKM-UQ from UCSC Xena Browser](https://xenabrowser.net/datapages/?dataset=TARGET-AML.htseq_fpkm-uq.tsv&host=https%3A%2F%2Fgdc.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)

```{r}
gene_TARGET <- read.csv("../Xena browser/TARGET-AML.htseq_fpkm-uq.tsv", 
                      sep="\t", check.names = FALSE, row.names = 1)
rownames(gene_TARGET) <- sub("\\..*$", "", rownames(gene_TARGET))

# merge samples of the same patient
fn_mean <- function(x) {
  x <- x[x>0]
  if (length(x) ==0) {
    return(0)
  } else if (length(x)==1) {
    return(x)
  } else {
    return(mean(x))
  }
}

patient_id <- sub("-.{3}$", "", colnames(gene_TARGET))
for (i in unique(patient_id)) {
  ind <- which(patient_id == i)
  if (length(ind) == 1) next
  gene_TARGET[,1] = apply(gene_TARGET[,ind], 1, fn_mean)
}
gene_TARGET <- gene_TARGET[,!duplicated(patient_id)]
colnames(gene_TARGET) <- unique(patient_id)

gene_TARGET[1:3,1:3]
```


```{r}
# Keep the 151 samples with RNA-seq from clinical data
table(clinical_TARGET$TARGET.USI %in% colnames(gene_TARGET))
table(clinical_TARGET2$Patient.ID %in% colnames(gene_TARGET))

clinical_TARGET <- clinical_TARGET[match(colnames(gene_TARGET), clinical_TARGET$TARGET.USI),]
clinical_TARGET2<- clinical_TARGET2[match(colnames(gene_TARGET), clinical_TARGET2$Patient.ID),]
```

# Variables of interest

## Adult AML (TCGA)

```{r}
df_TCGA <- data.frame(
  id = clinical_TCGA$submitter_id,
  time_to_death = clinical_TCGA$days_to_death.demographic / 30, # in months
  time_to_followup = clinical_TCGA$days_to_last_follow_up.diagnoses / 30, # in months
  vital_status = as.numeric(clinical_TCGA$vital_status.demographic == "Dead"),
  sex = clinical_TCGA$gender.demographic,
  age = round(clinical_TCGA$age_at_diagnosis.diagnoses / 365.25, 1), # in years
  FAB = clinical_TCGA$leukemia_french_american_british_morphology_code, # AML subtypes
  FLT3_ITD = clinical_TCGA$FLT3_ITD,
  cytogenetic_risk = clinical_TCGA$acute_myeloid_leukemia_calgb_cytogenetics_risk_category,
  cytogenetic_code = clinical_TCGA2$Cytogenetic.Code..Other., # for excluding BCR-ABL1, PML-RARA
  protocol = clinical_TCGA2$Induction, # chemotherapy
  transplant_type = clinical_TCGA2$Transplant.Type, # bone marrow transplant
  PC = as.numeric(gene_TCGA["ENSG00000173599",]),
  MPC1 = as.numeric(gene_TCGA["ENSG00000060762",]),
  MPC2 = as.numeric(gene_TCGA["ENSG00000143158",]),
  RHOF = as.numeric(gene_TCGA["ENSG00000139725",]),
  VIM = as.numeric(gene_TCGA["ENSG00000026025",])
) %>%
  mutate(
    overall_time = ifelse(vital_status==1, time_to_death, time_to_followup),
    sex = relevel(factor(sex), "female"),
    FAB = relevel(factor(FAB), "M0 Undifferentiated"),
    cytogenetic_risk = relevel(factor(cytogenetic_risk), "Favorable"),
    transplant_type = case_when(
      transplant_type == "0" ~ "No",
      grepl("MUD", transplant_type) ~ "MUD",
      grepl("sib Allo", transplant_type) ~ "sib Allo",
      grepl("Haplo", transplant_type) ~ "sib Allo",
      grepl("Auto", transplant_type) ~ "Auto"
    ),
    transplant_type = relevel(factor(transplant_type), "No"),
    protocol = case_when(
      protocol == "no treatment" ~ "No",
      grepl("MUD", protocol) ~ "MUD",
      grepl("7\\+", protocol) ~ "7+3",
      TRUE ~ "Others"
    ),
    protocol = relevel(factor(protocol), "No"),
    MPC = (MPC1 + MPC2)/2
  ) %>%
  relocate(overall_time, .before = 2) %>%
  select(-c(time_to_death, time_to_followup))

head(df_TCGA)

write.csv(df_TCGA, row.names = FALSE, file = "Suvival_data_TCGA.csv")
```

## Paediatric AML (TARGET)

```{r}
df_TARGET <- data.frame(
  id = clinical_TARGET$TARGET.USI,
  overall_time = clinical_TARGET$Overall.Survival.Time.in.Days / 30, # in months
  vital_status = as.numeric(clinical_TARGET$Vital.Status == "Dead"),
  sex = clinical_TARGET$Gender,
  age = round(clinical_TARGET$Age.at.Diagnosis.in.Days / 365.25, 1), # in years
  FAB = clinical_TARGET$FAB.Category, # AML subtypes
  FLT3_ITD = clinical_TARGET$FLT3.ITD.positive.=="Yes",
  cytogenetic_risk = clinical_TARGET$Cytogenetic.Complexity,
  cytogenetic_code = clinical_TARGET$Cytogenetic.Code.Other,
  protocol = clinical_TARGET$Protocol,
  induction = clinical_TARGET$Bone.Marrow.Site.of.Relapse.Induction.Failure != "Not done", # chemotherapy
  transplant = clinical_TARGET$SCT.in.1st.CR=="Yes", # stem cell transplant
  PC = as.numeric(gene_TARGET["ENSG00000173599",]),
  MPC1 = as.numeric(gene_TARGET["ENSG00000060762",]),
  MPC2 = as.numeric(gene_TARGET["ENSG00000143158",]),
  RHOF = as.numeric(gene_TARGET["ENSG00000139725",]),
  VIM = as.numeric(gene_TARGET["ENSG00000026025",])
) %>%
  mutate(
    sex = relevel(factor(sex), "Female"),
    FLT3_ITD = relevel(factor(FLT3_ITD), "FALSE"),
    induction = relevel(factor(induction), "FALSE"),
    transplant = relevel(factor(transplant), "FALSE"),
    FAB = relevel(factor(FAB), "M0"),
    cytogenetic_risk = case_when(
      cytogenetic_risk<3 ~ "Low",
      cytogenetic_risk<5 ~ "Medium",
      cytogenetic_risk>=5 ~ "High"
    ),
    cytogenetic_risk = relevel(factor(cytogenetic_risk), "Low"),
    protocol = factor(protocol),
    MPC = (MPC1 + MPC2)/2
  )

head(df_TARGET)

write.csv(df_TARGET, row.names = FALSE, file = "Suvival_data_TARGET.csv")
```

# Survival analysis

```{r}
# Function for getting the most frequent value (statistic mode)
fn_most_common <- function(x) {
  table(x) %>%
    sort(decreasing = TRUE) %>%
    {names(.)[1]}
}
```

## Adult AML (TCGA)

```{r}
df_TCGA <- filter(df_TCGA, cytogenetic_risk != "") %>%
  mutate(cytogenetic_risk = as.character(cytogenetic_risk),
         cytogenetic_risk = case_when(
           cytogenetic_risk == "Favorable" ~ "Low",
           cytogenetic_risk == "Intermediate/Normal" ~ "Intermediate",
           cytogenetic_risk == "Poor" ~ "High"
         ),
         cytogenetic_risk = factor(cytogenetic_risk, levels = c("Low","Intermediate","High"))
         )
ind <- with(df_TCGA, which(
  overall_time > 0 &
  FAB != "Not Classified"
))
df <- df_TCGA[ind,] %>%
  mutate(FAB = ifelse(FAB=="M0 Undifferentiated", "M0", as.character(FAB)),
         FAB = relevel(factor(FAB), "M0"),
         protocol = droplevels(protocol)
         ) %>%
  na.omit()
dim(df)
head(df)

write.csv(df, file = "adult_PC_cytogenetic_raw.csv")
```

### PC

```{r}
mod_TCGA_PC <- coxph(Surv(overall_time, vital_status) ~ 
                         age + sex + FLT3_ITD + protocol + transplant_type +
                         PC * (FAB + cytogenetic_risk),
                  data =  df)
suppressWarnings({
  mod_TCGA_PC <- step(mod_TCGA_PC, trace = 0)
})

summary(mod_TCGA_PC)
```

#### Plot (Cytogenetic risk)

```{r}
# Function to extract the source data for survival figures. The return table is formatted as time, survival status of group1 and group2.
extract_plot_data <- function(dat, outname = "") {
  ind1 = which(dat$n.event > 0)
  ind2 = which(dat$n.censor > 0)
  surv_data = rbind(
    data.frame(
      Months = rep(dat$time[ind1], times=dat$n.event[ind1]),
      Low = ifelse(rep(dat$strata[ind1], times=dat$n.event[ind1]) == "Low", as.character(dat$surv[ind1]), ""),
      High = ifelse(rep(dat$strata[ind1], times=dat$n.event[ind1]) == "High", as.character(dat$surv[ind1]), "")
    ),
    data.frame(
      Months = rep(dat$time[ind2], times=dat$n.censor[ind2]),
      Low = ifelse(rep(dat$strata[ind2], times=dat$n.censor[ind2]) == "Low", as.character(dat$surv[ind2]), ""),
      High = ifelse(rep(dat$strata[ind2], times=dat$n.censor[ind2]) == "High", as.character(dat$surv[ind2]), "")
    )
  )
  surv_data = surv_data[order(surv_data$Low=="", surv_data$Months),]
  colnames(surv_data) = c("Months", "Low PC expression", "High PC expression")
  rownames(surv_data) = NULL
  
  write.csv(surv_data, file = outname, row.names = F)
}
```


```{r fig.height=4, fig.height=4}
# p-value of the interaction term between PC and cytogenetic levels
pval_ = summary(mod_TCGA_PC)$coef[c("PC", "PC:cytogenetic_riskIntermediate", "PC:cytogenetic_riskHigh"),5] %>% round(3)

plot_list = list()
for (i in seq(length(levels(df$cytogenetic_risk)))) {
  plot_data = with(df, expand.grid(
    age = median(age),
    PC = quantile(PC, c(0.2, 0.8)),
    sex = levels(sex)[1],
    FAB = fn_most_common(FAB),
    cytogenetic_risk = levels(cytogenetic_risk)[i],
    FLT3_ITD = fn_most_common(FLT3_ITD),
    protocol = fn_most_common(protocol),
    transplant_type = fn_most_common(transplant_type)
  ))
  
  fit <- survfit(mod_TCGA_PC, newdata = plot_data, data=plot_data)
  plot_list[[i]] <- ggsurvplot(fit, 
             conf.int = T, 
             legend.labs=c("Low", "High"),
             title = table(df$cytogenetic_risk)[i] %>% {paste0(names(.)," (n=",.,")")},
             legend = c(0.8, 0.8),
             legend.title = "PC",
             xlab = "Follow-up time (month)",
             pval = pval_[i],
             pval.coord = c(ifelse(i==3,20,0),0.2),
             palette = rev(scales::hue_pal()(2)),
             ggtheme = theme_minimal())$plot +
  theme(plot.title = element_text(hjust=0.5),
        text = element_text(size = 15)
        ) +
  coord_fixed(ratio = 100)
}

plot_list[[1]]
plot_list[[2]]
plot_list[[3]]
```

```{r}
# Save plots
for (i in seq(length(plot_list))) {
  ggsave(filename = paste0("adult_PC_cytogenetic_", c("Low", "Intermediate", "High")[i], ".tiff"), 
         device = "tiff", dpi = 300, plot = plot_list[[i]], bg = "white", height = 4, width = 4)
  
  extract_plot_data(plot_list[[i]]$data, paste0("adult_PC_cytogenetic_", c("Low", "Intermediate", "High")[i], "_data.csv"))
}
```

#### exclude M3

```{r}
tem <- filter(df, FAB != "M3")
mod_TCGA_PC_3 <- coxph(Surv(overall_time, vital_status) ~ 
                         age + sex + FLT3_ITD + protocol + transplant_type +
                         PC * (FAB + cytogenetic_risk),
                  data =  tem)
suppressWarnings({
  mod_TCGA_PC_3 <- step(mod_TCGA_PC_3, trace = 0)
})

summary(mod_TCGA_PC_3)

write.csv(tem, file = "adult_excludeM3_PC_cytogenetic_raw.csv")
```

##### Plot (Cytogenetic risk)

```{r fig.height=4, fig.width=4}
pval_ = summary(mod_TCGA_PC_3)$coef[c("PC", "PC:cytogenetic_riskIntermediate", "PC:cytogenetic_riskHigh"),5] %>% round(3)

plot_list = list()
for (i in seq(length(levels(df$cytogenetic_risk)))) {
  plot_data = with(df, expand.grid(
    age = median(age),
    PC = quantile(PC, c(0.2, 0.8)),
    sex = levels(sex)[1],
    FAB = fn_most_common(FAB),
    cytogenetic_risk = levels(cytogenetic_risk)[i],
    FLT3_ITD = fn_most_common(FLT3_ITD),
    protocol = fn_most_common(protocol),
    transplant_type = fn_most_common(transplant_type)
  ))
  
  fit <- survfit(mod_TCGA_PC_3, newdata = plot_data, data=plot_data)
  plot_list[[i]] <- ggsurvplot(fit, 
             conf.int = T, 
             legend.labs=c("Low", "High"),
             title = table(tem$cytogenetic_risk)[i] %>% {paste0(names(.)," (n=",.,")")},
             legend = c(0.8, 0.8),
             legend.title = "PC",
             xlab = "Follow-up time (month)",
             pval = pval_[i],
             pval.coord = c(ifelse(i==3,20,0),0.2),
             palette = rev(scales::hue_pal()(2)),
             ggtheme = theme_minimal())$plot + 
    theme(plot.title = element_text(hjust=0.5),
        text = element_text(size = 15)
        ) +
  coord_fixed(ratio = 100)
}

plot_list[[1]]
plot_list[[2]]
plot_list[[3]]
```

```{r}
# Save plots
for (i in seq(length(plot_list))) {
  ggsave(filename = paste0("adult_excludeM3_PC_cytogenetic_", c("Low", "Intermediate", "High")[i], ".tiff"), 
         device = "tiff", dpi = 300, plot = plot_list[[i]], bg = "white", width = 4, height = 4)
 
  extract_plot_data(plot_list[[i]]$data, paste0("adult_excludeM3_PC_cytogenetic_", c("Low", "Intermediate", "High")[i], "_data.csv")) 
}
```

### MPC

```{r}
mod_TCGA_MPC <- coxph(Surv(overall_time, vital_status) ~ 
                         age + sex + FLT3_ITD + protocol + transplant_type +
                         MPC * (FAB + cytogenetic_risk),
                  data =  df)
suppressWarnings({
  mod_TCGA_MPC <- step(mod_TCGA_MPC, trace = 0)
})

summary(mod_TCGA_MPC)
```

## Paediatric AML (TARGET)

```{r}
df_TARGET <- df_TARGET %>%
  mutate(
    cytogenetic_risk = as.character(cytogenetic_risk),
    cytogenetic_risk = case_when(
      cytogenetic_risk=="Medium" ~ "Intermediate",
      TRUE ~ cytogenetic_risk
    ),
    cytogenetic_risk = factor(cytogenetic_risk, levels = c("Low","Intermediate","High"))
  )
ind <- which(
  df_TARGET$FAB != "Unknown" &
  df_TARGET$FAB != "NOS"
)
df2 <- df_TARGET[ind,] %>%
  mutate(FAB = ifelse(FAB=="M0 Undifferentiated", "M0", as.character(FAB)),
         FAB = relevel(factor(FAB), "M0")
         ) %>%
  na.omit()
dim(df2)
head(df2)

```

### PC

```{r}
mod_TARGET_PC <- coxph(Surv(overall_time, vital_status) ~
                         age + sex + FLT3_ITD + protocol + induction + transplant +
                         PC * (FAB + cytogenetic_risk),
                  data =  df2)
suppressWarnings({
  mod_TARGET_PC <- step(mod_TARGET_PC, trace=0)
})
summary(mod_TARGET_PC)
```

#### Plot (Cytogenetic risk)

```{r fig.height=4, fig.width=8}
pval_ = summary(mod_TARGET_PC)$coef[c("PC","PC:cytogenetic_riskIntermediate","PC:cytogenetic_riskHigh"),5] %>% round(3)
pval_[is.na(pval_)] = "NS"

plot_list = list()
for (i in seq(length(levels(df2$cytogenetic_risk)))) {
  plot_data = with(df2, expand.grid(
    age = median(age),
    PC = quantile(PC, c(0.2, 0.8)),
    sex = levels(sex)[1],
    FAB = fn_most_common(FAB),
    cytogenetic_risk = levels(cytogenetic_risk)[i],
    FLT3_ITD = fn_most_common(FLT3_ITD),
    protocol = fn_most_common(protocol),
    induction = fn_most_common(induction),
    transplant = fn_most_common(transplant)
  ))
  
  fit <- survfit(mod_TARGET_PC, newdata = plot_data, data=plot_data)
  plot_list[[i]] <- ggsurvplot(fit, 
             conf.int = T, 
             legend.labs=c("Low", "High"),
             title = table(df2$cytogenetic_risk)[i] %>% {paste0(names(.)," (n=",.,")")},
             legend = c(0.8, 0.8),
             legend.title = "PC",
             xlab = "Follow-up time (month)",
             pval = pval_[i],
             ggtheme = theme_minimal())$plot + theme(plot.title = element_text(hjust=0.5))
}

cowplot::plot_grid(plotlist = plot_list,
                   nrow=2)
```

#### Plot (FAB)

```{r fig.height=6, fig.width=12}
pval_ = summary(mod_TARGET_PC)$coef[c("PC",paste0("PC:FABM", c(1,2,4:7))),5] %>% round(3)
pval_[is.na(pval_)] = "NS"

plot_list = list()
for (i in seq(length(levels(df2$FAB)))) {
  plot_data = with(df2, expand.grid(
    age = median(age),
    PC = quantile(PC, c(0.2, 0.8)),
    sex = levels(sex)[1],
    FAB = levels(FAB)[i],
    cytogenetic_risk = fn_most_common(cytogenetic_risk),
    FLT3_ITD = fn_most_common(FLT3_ITD),
    protocol = fn_most_common(protocol),
    induction = fn_most_common(induction),
    transplant = fn_most_common(transplant)
  ))
  
  fit <- survfit(mod_TARGET_PC, newdata = plot_data, data=plot_data)
  plot_list[[i]] <- ggsurvplot(fit, 
             conf.int = T, 
             legend.labs=c("Low", "High"),
             title = table(df2$FAB)[i] %>% {paste0(names(.)," (n=",.,")")},
             legend = c(0.8, 0.8),
             legend.title = "PC",
             xlab = "Follow-up time (month)",
             pval = pval_[i],
             ggtheme = theme_minimal())$plot + theme(plot.title = element_text(hjust=0.5))
}

cowplot::plot_grid(plotlist = plot_list,
                   nrow=3)
```

### MPC

```{r}
mod_TARGET_MPC <- coxph(Surv(overall_time, vital_status) ~
                         age + sex + FLT3_ITD + protocol + induction + transplant +
                         MPC * (FAB + cytogenetic_risk),
                  data =  df2)
suppressWarnings({
  mod_TARGET_MPC <- step(mod_TARGET_MPC, trace=0)
})
summary(mod_TARGET_MPC)
```

#### Plot (MPC)

```{r}
pval_ = summary(mod_TARGET_MPC)$coef[c("MPC"),5] %>% round(3)
pval_[is.na(pval_)] = "NS"

plot_data = with(df2, expand.grid(
  age = median(age),
  MPC = quantile(MPC, c(0.2, 0.8)),
  sex = levels(sex)[1],
  FAB = fn_most_common(FAB),
  cytogenetic_risk = fn_most_common(cytogenetic_risk),
  FLT3_ITD = fn_most_common(FLT3_ITD),
  protocol = fn_most_common(protocol),
  induction = fn_most_common(induction),
  transplant = fn_most_common(transplant)
))

fit <- survfit(mod_TARGET_MPC, newdata = plot_data, data=plot_data)
ggsurvplot(fit, 
           conf.int = T, 
           legend.labs=c("Low", "High"),
           legend = c(0.8, 0.8),
           legend.title = "MPC",
           xlab = "Follow-up time (month)",
           pval = pval_,
           ggtheme = theme_minimal())$plot + theme(plot.title = element_text(hjust=0.5))

```

# Session info

```{r}
sessionInfo()
```


